// pixelmatch/max.js
'use strict';

exports.rgba = rgbaDist;
exports.rgb = rgbDist;

exports.rgba2y = rgba2y;
exports.rgb2y = rgb2y;

function rgbDist(r1, g1, b1, r2, g2, b2) {
    var y = rgb2y(r1, g1, b1) - rgb2y(r2, g2, b2),
        i = rgb2i(r1, g1, b1) - rgb2i(r2, g2, b2),
        q = rgb2q(r1, g1, b1) - rgb2q(r2, g2, b2);

    return y * y * 0.5053 + i * i * 0.299 + q * q * 0.1957;
}

function rgbaDist(r1, g1, b1, a1, r2, g2, b2, a2, br, bg, bb) {
    return rgbDist(
        br + (r1 - br) * a1,
        bg + (g1 - bg) * a1,
        bb + (b1 - bb) * a1,
        br + (r2 - br) * a2,
        bg + (g2 - bg) * a2,
        bb + (b2 - bb) * a2);
}

function rgba2y(r, g, b, a, br, bg, bb) {
    return rgb2y(
        br + (r - br) * a,
        bg + (g - bg) * a,
        bb + (b - bb) * a);
}

function rgb2y(r, g, b) { return r * 0.29889531 + g * 0.58662247 + b * 0.11448223; }
function rgb2i(r, g, b) { return r * 0.59597799 - g * 0.27417610 - b * 0.32180189; }
function rgb2q(r, g, b) { return r * 0.21147017 - g * 0.52261711 + b * 0.31114694; }
